### PR TITLE
Improve the documentation for std::convert (From, Into, AsRef and AsMut)

### LINKED ISSUE #59163: Make the documentation of From and Into traits coherent.
# Description
I was reading on the From<T> and Into<T> traits and I got a little confused:

```From``` states: 
> **Simple and safe** type conversions in to Self.

whereas ```Into``` states:
> A conversion that consumes self, **which may or may not be expensive**.

### The Problem
This makes it seem like implementing ```From``` should be used for implementing relatively cheap operations (interpreting the word 'simple' as 'cheap') automatically gaining ```Into``` for free. Whereas implementing ```Into``` (and therefore not getting the ```From``` trait for free) should be used for expensive conversions.

However later in the documentation of ```Into``` states:
> Library authors should not directly implement this trait, but should prefer implementing the From trait, which offers greater flexibility and provides an equivalent Into implementation for free, thanks to a blanket implementation in the standard library.

which in turns suggests that one should not implement ```Into``` at all (unless you want to convert into a type outside your crate).

Now I have 3 questions:

1. Is it true, that ```From``` should only be used for inexpensive conversions? (If not, we should probably change the documentation of both ```From``` and ```Into``` to refrain from saying things about expensiveness).

2. The documentation of ```Into``` states that one should only implement ```Into``` if you want to convert into a type outside your crate. ```From``` cannot do this as the example shows. However the significant limitation is that you do not get the ```From``` trait for free, meaning that library authors using ```From``` as trait bounds cannot automatically use the ```Into``` implementation. Is there a reason why implementing ```Into``` does not automatically implement ```From``` as well?

3. ```Into``` nicely states that it should only be used when implementing traits for types in external crates. Maybe the documentation of ```From``` should also note this?

I am curious on what you all think!

**Disclaimer:** I have been using Rust for the last 4 months, so I would not consider myself an expert. So if I am wrong on some points or misunderstand anything, please correct me!

### PR DESCRIPTION
**Pull request by DevQps**
# Description
In this PR I updated the documentation of From, Into, AsRef and AsMut, as well as the general std::convert module documentation. The discussion in #59163 provided information that was not yet present in the docs, or was not expressed clearly enough. I tried to clarify the examples that were already present in the docs as well as add more information about considered best-practices that came out of the discussion in #59163 

@steveklabnik I hope I didn't change too much. This is an initial version! I will scan through everything tomorrow as well again to see if I made any typo's or errors, and maybe make some small changes here and there.

All suggestions are welcome!

closes #59163

### CODE REVIEW

File: src/libcore/convert.rs
Comment (lzutao) [inline src/libcore/convert.rs]:
```rust
//! - Implement the [`Into`] trait for consuming value-to-value conversions to types
//!   outside
```

File: src/libcore/convert.rs
Comment (lzutao) [inline src/libcore/convert.rs]: similar

**Review by tesuji — COMMENTED** [review]
Some nits on indentations.

File: src/libcore/convert.rs
Comment (DevQps) [reply src/libcore/convert.rs]: @lu-zero Thanks for your response! I will go and fix this (Y)

Comment (DevQps) [reply src/libcore/convert.rs]: Should be fixed now!

**Comment by steveklabnik** [comment]
Hey hey! I won't have time until tomorrow to review this; @rust-lang/docs can you check it out in the meantime?

Thanks so much!

File: src/libcore/convert.rs
Comment (ehuss) [inline src/libcore/convert.rs]: I think the original link here is the correct one.  The first-edition is no longer included.

File: src/libcore/convert.rs
Comment (lzutao) [inline src/libcore/convert.rs]: Why use \`\`\` instead of singular \' for inlining code?

File: src/libcore/convert.rs
Comment (DevQps) [reply src/libcore/convert.rs]: Thanks for pinpointing this. I apparently have made an error, because it indeed seems like the removed line contained the correct reference. I will change it right away!

File: src/libcore/convert.rs
Comment (DevQps) [reply src/libcore/convert.rs]: To be truly honest, I never knew singular ` would work as well. I just changed all the code that is inlined in the text using singular ` as you proposed.

Thanks for your suggestion!

Comment (lzutao) [reply src/libcore/convert.rs]: You're welcome! A small nits: \` called [backtick or backquote](https://unix.stackexchange.com/questions/27428/what-does-backquote-backtick-mean-in-commands).

**Comment by lzutao** [comment]
One of the thing I find that it's hard to review docs PR because we don't have 
visual built documentation of the changes right away.

File: src/libcore/convert.rs
Comment (DevQps) [reply src/libcore/convert.rs]: Thanks! I didn't know that either :) I'll keep it in mind for the future. I guess there's still always a lot to learn out there :)

**Comment by DevQps** [comment]
@lzutao I can provide a tar or zip of the generated docs somehoe but that isn't really ideal of course. Maybe it can be added as a downloadable asset at each Travis CI pipeline? I used to do that with Gitlab pipelines, not sure if Travis can do such a thing as well though.

**Comment by DevQps** [comment]
@ehuss @lzutao @steveklabnik Let me know if there's anything else I can do to improve things!

**Comment by lzutao** [comment]
This PR needs an review from [Documentation team](https://www.rust-lang.org/governance/teams/documentation) I think. You could ping any of them for requesting review.

**Comment by steveklabnik** [comment]
I'm planning on reviewing within the next few hours.

File: src/libcore/convert.rs
Comment (steveklabnik) [inline src/libcore/convert.rs]: So, I can appreciate that elaborating here makes sense, but it doesn't fit in with the rest of the style of the docs. These summary lines appear in places like here: 

image

Note that most of these summaries follow this style; there are a few which are more like what you've written, but those should probably be fixed to fit this style.

Anyway, what I'm saying is, let's leave the original summary here. :)

File: src/libcore/convert.rs
Comment (steveklabnik) [inline src/libcore/convert.rs]: ... however, we *can* move it to be part of this paragraph though; I think that's where it fits in.

File: src/libcore/convert.rs
Comment (steveklabnik) [inline src/libcore/convert.rs]: There needs to be a `///` between this line and the next one; the format is

```rust
/// summary line
///
/// full description
```

the previous doc block didn't follow this closely, so it's not your fault! but it's a thing we should fix. :)

File: src/libcore/convert.rs
Comment (steveklabnik) [inline src/libcore/convert.rs]: putting another `///` before this line to give the list some room to breathe would be great.

File: src/libcore/convert.rs
Comment (steveklabnik) [inline src/libcore/convert.rs]: there needs to be a `///` above this line for this to render as two paragraphs

File: src/libcore/convert.rs
Comment (steveklabnik) [inline src/libcore/convert.rs]: same issue with the summary here

File: src/libcore/convert.rs
Comment (steveklabnik) [inline src/libcore/convert.rs]: if you wanted this to be a separate paragraph, this needs a `///` above it. howevr, I'm not sure it needs to be one, in which case, it should be better wrapped with the above line

File: src/libcore/convert.rs
Comment (steveklabnik) [inline src/libcore/convert.rs]: please put a `///` above this line to give some space between the heading and the body :)

File: src/libcore/convert.rs
Comment (steveklabnik) [inline src/libcore/convert.rs]: same here, `///` above please!

File: src/libcore/convert.rs
Comment (steveklabnik) [inline src/libcore/convert.rs]: and here

File: src/libcore/convert.rs
Comment (steveklabnik) [inline src/libcore/convert.rs]: This paragraph also has some weird wrapping, could you fix that, please?

File: src/libcore/convert.rs
Comment (steveklabnik) [inline src/libcore/convert.rs]: same here

**Review by steveklabnik — CHANGES_REQUESTED** [review]
Hey hey! This PR is *great* content-wise, I just have some formatting stuff. Please let me know if any of that isn't clear! As soon as the formatting is fixed, we can merge this in 👍

**Comment by DevQps** [comment]
> Hey hey! This PR is _great_ content-wise, I just have some formatting stuff. Please let me know if any of that isn't clear! As soon as the formatting is fixed, we can merge this in 👍

@steveklabnik Thanks for your kind words! I fixed all the issues you brought up, thanks a lot for all the feedback :) Many of the things I was unaware of / didn't know so it will probably help me when I do some editing in the future as well.

I was wondering!: Are there are other things I could maybe help with? I am not sure if there are still documents to be reviewed? Or if there is some documentation that needs a test read / that needs some improvement? Now that I got the hang of it a little bit I would enjoy contributing some more! Let me know if there's anything I can do to help :)

**Review by steveklabnik — APPROVED** [review]

**Comment by steveklabnik** [comment]
Awesome, thanks! 

@bors: r+ rollup

> I was wondering!: Are there are other things I could maybe help with? I am not sure if there are still documents to be reviewed? Or if there is some documentation that needs a test read / that needs some improvement? Now that I got the hang of it a little bit I would enjoy contributing some more! Let me know if there's anything I can do to help :)

For sure! We have a T-docs tag with issues. Picking out one of them and writing some docs would be fantastic!

**Pull request merged**
